Strategy from paper

rank-and-file opened the pull request:
Basic strategy from the paper

Commit:
Added badHintSubroutine and the twoStepsOrthogonal method

Commit:
switch case in twoStepsOrthogonal and correct the good hint condition

Commit:
Go to center of the rectangle after incrementing the phase

Commit:
Init test class StrategyFromPaperTest

Commit:
Init methods for twoHintsSubroutine and add TODOs

Commit:
Merge branch 'master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java
#	src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java

Commit:
Resolve conflicts that emerged through the merge

Commit:
Some renamings

Commit:
Add thoughts of strategy to the return value of the move method

Commit:
Add test for basic move method and debug rectangleScan

Commit:
Changed HalfPlaneHint to implementation that i think is the most
appropriate

Commit:
Change impl of HalfPlaneHint once again and add impls rho sigma phi reversePhi and getBasicTransformation in StrategyFromPaper

Commit:
Refactored lineLineSegmentIntersection to lineWayIntersection

Commit:
Add variable initialisation in twoHintSubroutine

Commit:
Changes in StrategyFromPaper to fit the new lineWayIntersection method

Commit:
Add lastHintBadSubroutine and rewrite many old methods

The adding of lastHintBadSubroutine has shown that some methods aren't
practiacal to use how they were defined and therefore I rewrote them.

Commit:
Add some comments

Commit:
Add some tests

Commit:
Add author and remove rhoTest

Commit:
Add some comments in JTSUtils

Commit:
Add some comments in StrategyFromPaper

Commit:
Merge remote-tracking branch 'origin/master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java
#	src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java
#	src/main/java/com/treasure/hunt/utils/JTSUtils.java
#	src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java

Commit:
Rewrite one comment in JTSUtils

Commit:
Add RandomHalfPlaneHintHider

Commit:
Add getGeometryItems to HalfPlaneHint

Commit:
Add StrategyFromPaperAngleHintWrapper and fix some bugs in
StrategyFromPaper

Commit:
Bugfixes

Commit:
Merge branch 'master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/strategy/geom/GeometryType.java

Commit:
Debug some methods in the strategy (incl lastHintBadSubroutine,
arrangeRectangle) and add tests

Commit:
Changes in arragneRectangle and manually create matrices in sigmaPoint

Commit:
Rotate around r in sigmaPoint and add the center of the rectangle to
move in lastHintBadSubroutine

Commit:
Move arrangeRectangle in phiOtherRetangleInverse

Commit:
Increase scale of PrecisionModel

Commit:
Merge branch 'master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java
#	src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java
#	src/main/java/com/treasure/hunt/utils/JTSUtils.java

Commit:
Change Coordiante to Point in HalfPlaneHint

Commit:
Use manual matrix in sigmaRectangle instead of generated one

Commit:
Add the lines of the way in moveReturn

Rename setLocationAndReturn to moveReturn and add the lines which are
described by the parameter move to the return value.

Commit:
Redo visualisations for strategyFromPaper

In GeometryStyle change colours
In GeometryType change override options
In HalfPlaneHint change Coordinate to Point (was a bug)

Commit:
Redo randomness in RandomHalfPlaneHintHider

RandomHalfPlaneHintHider now also can spawn the treasure in other
quadrants (before the change it could only spawn them in the first
quadrant)

Commit:
Change the construction of HalfPlaneHints in the test

Commit:
Rename points of and add visualization of HalfPlaneHint

Rename points to leftPoint and rightPoint (before they were named
anglePointRight and anglePointLeft)
A HalfPlaneHint is now visualized through a line.

Commit:
Make Playing area bigger and fix visualization bug in HalfPlaneHint

Commit:
Init routinesFromPaper and rectangleUtils

Commit:
Init BadHintSubroutine and Test in strategyFromPaper

Commit:
Add comments

Commit:
Merge remote-tracking branch 'origin/master' into strategyFromPaperUndo

# Conflicts:
#	src/main/java/com/treasure/hunt/game/GameEngine.java
#	src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java

Commit:
Change visualisation of HalfPlaneHints, they now are lines

Commit:
Renamed strategyFromPaper.Test to strategyFromPaper.Tester

Commit:
Restructure code in BadHintSubroutine and undo one bug in
RoutinesFromPaper

Commit:
Add test for phiRect and phiRectReverse

Commit:
Try to color last move in another color (doesn't work though)

Commit:
Remove some nonused GeometryTypes

Commit:
Merge branch 'master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/game/GameEngine.java
#	src/main/java/com/treasure/hunt/jts/awt/HalfPlane.java
#	src/main/java/com/treasure/hunt/strategy/geom/GeometryType.java
#	src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java
#	src/main/java/com/treasure/hunt/utils/JTSUtils.java

Commit:
Change GeometryTypes to fit current masters implementation

Commit:
Remove System.out.println()s and change some comments

Commit:
Delete some useless tests and renaming

Commit:
Delete StrategyUtils (empty class)

Commit:
Redo init to match current interface

Commit:
Let HalfPlaneHint inherit from AngleHint

Commit:
Change to fit AngleHints constructor (there was a bug)

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/hint/impl/HalfPlaneHint.java:
```diff
@@ -1,107 +1,217 @@
     @Getter
-    private Direction direction;    // when the line indicated by halfplanePointOne and halfplanePointTwo is not horizontal,
+    private Direction direction;
+    // when the line indicated by anglePointLeft and anglePointRight is not horizontal,
```
This should be java doc

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+            throw ee;
+        }
+
+    }
```
This method is way to long split it up

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+     * @return The move to scan various areas so that A,B,C and D can be updated to a smaller rectangle (or the treasure
+     * is found)
+     */
+    static Movement lastHintBadSubroutine(StrategyFromPaper strategy, HalfPlaneHint curHint, HalfPlaneHint lastBadHint,
```
Why is everything static , maybe make it a singleton or well u can also leave it

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+        System.out.println(" D= (" + D.getX() + ", " + D.getY() + ")");
+        for (int i = 0; i < rect.length; i++)
+            System.out.println("rect[" + i + "]= " + rect[i]);
+        System.out.println("lastBadHint p1= " + lastBadHint.getCenter() + "lastHint p2= " +
```
1 Finger

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+        System.out.println("A= (" + A.getX() + ", " + A.getY() + ")");
+        System.out.println(" B= (" + B.getX() + ", " + B.getY() + ")");
+        System.out.println(" C= (" + C.getX() + ", " + C.getY() + ")");
+        System.out.println(" D= (" + D.getX() + ", " + D.getY() + ")");
```
Ganz viele Finger

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+            System.out.println("rect[" + i + "]= " + rect[i]);
+        System.out.println("lastBadHint p1= " + lastBadHint.getCenter() + "lastHint p2= " +
+                lastBadHint.getRight());
+        System.out.println("curHint p1= " + curHint.getCenter() + "curHint p2= " +
```
Noch mehr Fnger

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java:
```diff
@@ -0,0 +1,110 @@
+            max_y = Math.max(max_y, y);
+            min_x = Math.min(min_x, x);
+            min_y = Math.min(min_y, y);
+
```
newline too much

1-alex98 commented on src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java:
```diff
@@ -0,0 +1,114 @@
+
+        hint = new HalfPlaneHint(new Coordinate(-4, -0.5), new Coordinate(4, -0.5), up);
+        points = strat.move(hint).getPoints();
+        true_points = new Point[]{createPoint(0.75, -0.5), createPoint(0.75, 1.25)};
```
Stick to variable naming conventions

1-alex98 commented on src/test/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaperTest.java:
```diff
@@ -0,0 +1,114 @@
+        for (int i = 0; i < points.size(); i++) {
+            System.out.print(points.get(i).getObject());
+        }
+        System.out.println();
```
Finger

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java:
```diff
@@ -0,0 +1,336 @@
+        Point[] a = new Point[k + 1];
+        Point[] b = new Point[k + 1];
+
+        { //create a_i on line segment AB
```
Well whatever this is supposed to be probably extract it into a method

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/RoutinesFromPaper.java:
```diff
@@ -0,0 +1,336 @@
+        }
+
+        if (k % 2 == 1) //code like in paper
+        {
```
Tick the format boxes in the commit message or str+alt+l { should be same lie as loop

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java:
```diff
@@ -0,0 +1,110 @@
+        assertRectangle(rect);
+
+        Coordinate A = null, B = null, C = null, D = null;
+        double max_x = -Double.MAX_VALUE;
```
var naming conventions

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java:
```diff
@@ -0,0 +1,110 @@
+    static Coordinate[] arrangeRectangle(Coordinate[] rect) {
+        assertRectangle(rect);
+
+        Coordinate A = null, B = null, C = null, D = null;
```
What the hack are A B C And D

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/GeometricUtils.java:
```diff
@@ -0,0 +1,110 @@
+public class GeometricUtils {
+
+    static void assertRectangle(Coordinate[] rect) {
+        if (rect.length != 4)
```
Then lets call it reactangle

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+        System.out.println("lastBadHint p1= " + lastBadHint.getCenter() + "lastHint p2= " +
+                lastBadHint.getRight());
+        System.out.println("curHint p1= " + curHint.getCenter() + "curHint p2= " +
+                curHint.getRight());
```
Abbreviations are bad for your mental health

1-alex98 commented on src/main/java/com/treasure/hunt/strategy/searcher/impl/strategyFromPaper/BadHintSubroutine.java:
```diff
@@ -0,0 +1,227 @@
+                newRectangle = phiOtherRectangleInverse(basicTrans, rect,
+                        new Coordinate[]{At, g, p, m});
+            }
+            if ((x2_apos == left &&
```
Maybe a more descriptive name also namingconventions

1-alex98 reviewed (changes requested):
At least remove al System.outs

Commit:
Undo changes to moveReturn

Commit:
Rename some variables

Commit:
Remove System.out.println()

Commit:
Add javadoc and extract one thing out of rectangleScan into a method

Commit:
In GeometricUtils.arrangeRectangle rename A to newA, B to newB, ...

Commit:
Description of direction now is javadoc

Commit:
Add description for max_x, min_x, etc.

Commit:
Rename testedHint to testHint

Commit:
Delete useless line in StrategyFromPaperTest

Commit:
Rename x_max to xMax, x_min to xMin, etc

1-alex98 commented:
Still not all old comments addressed at least address System.out.println

rank-and-file commented:
> Still not all old comments addressed at least address System.out.println

Yes, I'm going to change that

Commit:
Replace System.out.printlns with error message in BadHintSubroutine

Commit:
Remove underscores from variablenames in lastHintBadSubroutine

Commit:
Change access modifiers and add private constructors

to BadHintSubroutine GeometricUtils and RoutinesFromPaper

Commit:
Make BadHintSubroutine Singleton

Commit:
Split the lastHintBadSubroutine method

Commit:
Rename A, B, C and D in StrategyFromPaper in more fitting names

Commit:
Fixed Typo

Commit:
Merge branch 'master' into strategyFromPaper

# Conflicts:
#	src/main/java/com/treasure/hunt/strategy/geom/GeometryStyle.java

Commit:
Add basic status messages to StrategyFromPaper

The status messages contain information about wether the hint was good
or bad and what the current basic transformation is if the last hint was
bad.

Commit:
Add some more stati (doesn't work as intended yet)

Commit:
Fix bug in RoutinesFromPaper

Somehow k was calculated two times in rectangleScan and the second time
sometimes got a different result

Commit:
Moved some things out of move method

Commit:
Remove another System.out.println

Commit:
Handle lastHintQuality different

Commit:
Change current to previous and last to before previous in status-messages

Commit:
Add description for rectangleScans and reduction of the rectangle

to the status, when the previous hint was bad
